File: python/advanced_ds/heavy_light_decomposition.py
"""
Heavy-Light Decomposition (HLD)

Time Complexity:
- Decomposition: O(n)
- Path Query/Update: O(log² n) with segment tree
- Point Update: O(log n) with segment tree

Space Complexity: O(n)

Use Cases:
- Answering min/max/sum queries on tree paths
- Updating values on tree paths
- LCA (Lowest Common Ancestor) queries with path info

Key Insight:
- Decompose tree into heavy paths (heavy child = child with most vertices in subtree)
- Each vertex has exactly one heavy child
- Tree has O(log n) heavy paths from root to any node
- Use segment tree to query along paths
"""

from typing import List, Optional, Tuple
from collections import defaultdict


class SegmentTree:
    """Segment tree for range min/max/sum queries."""

    def __init__(self, arr: List[int]):
        self.n = len(arr)
        self.tree = [0] * (4 * self.n)
        if self.n > 0:
            self._build(arr, 0, 0, self.n - 1)

    def _build(self, arr: List[int], node: int, start: int, end: int) -> None:
        """Build segment tree recursively."""
        if start == end:
            self.tree[node] = arr[start]
        else:
            mid = (start + end) // 2
            self._build(arr, 2 * node + 1, start, mid)
            self._build(arr, 2 * node + 2, mid + 1, end)
            self.tree[node] = max(self.tree[2 * node + 1], self.tree[2 * node + 2])

    def update(self, idx: int, val: int) -> None:
        """Update element at index idx to val."""
        self._update(0, 0, self.n - 1, idx, val)

    def _update(self, node: int, start: int, end: int, idx: int, val: int) -> None:
        if start == end:
            self.tree[node] = val
        else:
            mid = (start + end) // 2
            if idx <= mid:
                self._update(2 * node + 1, start, mid, idx, val)
            else:
                self._update(2 * node + 2, mid + 1, end, idx, val)
            self.tree[node] = max(self.tree[2 * node + 1], self.tree[2 * node + 2])

    def query(self, l: int, r: int) -> int:
        """Query max in range [l, r]."""
        if l > r or self.n == 0:
            return 0
        return self._query(0, 0, self.n - 1, l, r)

    def _query(self, node: int, start: int, end: int, l: int, r: int) -> int:
        if r < start or end < l:
            return 0
        if l <= start and end <= r:
            return self.tree[node]
        mid = (start + end) // 2
        left_max = self._query(2 * node + 1, start, mid, l, r)
        right_max = self._query(2 * node + 2, mid + 1, end, l, r)
        return max(left_max, right_max)


class HeavyLightDecomposition:
    """Heavy-Light Decomposition for tree path queries."""

    def __init__(self, n: int, edges: List[Tuple[int, int]], values: List[int]):
        """
        Initialize HLD.

        Args:
            n: Number of vertices (0 to n-1)
            edges: List of (u, v) edges (undirected)
            values: List of vertex values
        """
        self.n = n
        self.values = values.copy()
        self.adj: List[List[int]] = [[] for _ in range(n)]

        for u, v in edges:
            self.adj[u].append(v)
            self.adj[v].append(u)

        # HLD arrays
        self.parent = [-1] * n
        self.depth = [0] * n
        self.subtree_size = [0] * n
        self.heavy_child = [-1] * n
        self.chain_id = [-1] * n
        self.pos_in_chain = [-1] * n
        self.chain_head = [-1] * n

        self.chain_nodes: List[List[int]] = []
        self.seg_trees: List[SegmentTree] = []

        if n > 0:
            self._dfs1(0, -1, 0)
            self._dfs2(0, 0)

    def _dfs1(self, u: int, p: int, d: int) -> None:
        """First DFS: compute subtree sizes and find heavy children."""
        self.parent[u] = p
        self.depth[u] = d
        self.subtree_size[u] = 1

        for v in self.adj[u]:
            if v != p:
                self._dfs1(v, u, d + 1)
                self.subtree_size[u] += self.subtree_size[v]

                if self.heavy_child[u] == -1 or \
                   self.subtree_size[v] > self.subtree_size[self.heavy_child[u]]:
                    self.heavy_child[u] = v

    def _dfs2(self, u: int, chain_id: int) -> None:
        """Second DFS: assign chain IDs and positions."""
        self.chain_id[u] = chain_id

        if chain_id >= len(self.chain_nodes):
            self.chain_nodes.append([])
            self.chain_head.append(-1)

        if len(self.chain_nodes[chain_id]) == 0:
            self.chain_head[chain_id] = u

        self.pos_in_chain[u] = len(self.chain_nodes[chain_id])
        self.chain_nodes[chain_id].append(u)

        if self.heavy_child[u] != -1:
            self._dfs2(self.heavy_child[u], chain_id)

        for v in self.adj[u]:
            if v != self.parent[u] and v != self.heavy_child[u]:
                self._dfs2(v, len(self.chain_nodes))

    def _build_segment_trees(self) -> None:
        """Build segment trees for each chain."""
        self.seg_trees = []
        for chain in self.chain_nodes:
            chain_values = [self.values[u] for u in chain]
            self.seg_trees.append(SegmentTree(chain_values))

    def update(self, u: int, val: int) -> None:
        """Update value at vertex u."""
        self.values[u] = val
        self._build_segment_trees()

    def query_path(self, u: int, v: int) -> int:
        """Query max value on path from u to v."""
        if not self.seg_trees:
            self._build_segment_trees()

        result = 0

        # Bring u and v to same level
        while self.depth[u] > self.depth[v]:
            result = max(result, self._query_up(u))
            u = self.parent[u]
        while self.depth[v] > self.depth[u]:
            result = max(result, self._query_up(v))
            v = self.parent[v]

        # Move both up simultaneously
        while u != v:
            result = max(result, self._query_up(u), self._query_up(v))
            u = self.parent[u]
            v = self.parent[v]

        result = max(result, self.values[u])
        return result

    def _query_up(self, u: int) -> int:
        """Query from u to head of its chain."""
        chain_id = self.chain_id[u]
        start = self.pos_in_chain[self.chain_head[chain_id]]
        end = self.pos_in_chain[u]
        return self.seg_trees[chain_id].query(start, end)


if __name__ == "__main__":
    # Example: Tree with path queries
    #     0
    #    / \
    #   1   2
    #  / \
    # 3   4

    n = 5
    edges = [(0, 1), (0, 2), (1, 3), (1, 4)]
    values = [10, 20, 30, 40, 50]

    hld = HeavyLightDecomposition(n, edges, values)

    print("Path 3->4 max value:", hld.query_path(3, 4))  # 3->1->4, values: 40, 20, 50
    print("Path 3->2 max value:", hld.query_path(3, 2))  # 3->1->0->2, values: 40, 20, 10, 30
    print("Path 4->2 max value:", hld.query_path(4, 2))  # 4->1->0->2, values: 50, 20, 10, 30

    hld.update(1, 100)
    print("After update(1, 100):")
    print("Path 3->4 max value:", hld.query_path(3, 4))  # 40, 100, 50

File: java/advanced_ds/HeavyLightDecomposition.java
package advanced_ds;

import java.util.*;

/**
 * Heavy-Light Decomposition for tree path queries.
 *
 * Time Complexity:
 * - Decomposition: O(n)
 * - Path Query/Update: O(log² n) with segment tree
 * - Point Update: O(log n) with segment tree
 *
 * Space Complexity: O(n)
 *
 * Use Cases:
 * - Answering min/max/sum queries on tree paths
 * - Updating values on tree paths
 * - LCA (Lowest Common Ancestor) queries with path info
 *
 * Key Insight:
 * - Decompose tree into heavy paths (heavy child = child with most vertices in subtree)
 * - Each vertex has exactly one heavy child
 * - Tree has O(log n) heavy paths from root to any node
 * - Use segment tree to query along paths
 */
public class HeavyLightDecomposition {
    static class SegmentTree {
        int[] tree;
        int n;

        SegmentTree(int[] arr) {
            this.n = arr.length;
            this.tree = new int[4 * n];
            if (n > 0) {
                build(arr, 0, 0, n - 1);
            }
        }

        private void build(int[] arr, int node, int start, int end) {
            if (start == end) {
                tree[node] = arr[start];
            } else {
                int mid = (start + end) / 2;
                build(arr, 2 * node + 1, start, mid);
                build(arr, 2 * node + 2, mid + 1, end);
                tree[node] = Math.max(tree[2 * node + 1], tree[2 * node + 2]);
            }
        }

        void update(int idx, int val) {
            update(0, 0, n - 1, idx, val);
        }

        private void update(int node, int start, int end, int idx, int val) {
            if (start == end) {
                tree[node] = val;
            } else {
                int mid = (start + end) / 2;
                if (idx <= mid) {
                    update(2 * node + 1, start, mid, idx, val);
                } else {
                    update(2 * node + 2, mid + 1, end, idx, val);
                }
                tree[node] = Math.max(tree[2 * node + 1], tree[2 * node + 2]);
            }
        }

        int query(int l, int r) {
            if (l > r || n == 0) return 0;
            return query(0, 0, n - 1, l, r);
        }

        private int query(int node, int start, int end, int l, int r) {
            if (r < start || end < l) return 0;
            if (l <= start && end <= r) return tree[node];
            int mid = (start + end) / 2;
            int leftMax = query(2 * node + 1, start, mid, l, r);
            int rightMax = query(2 * node + 2, mid + 1, end, l, r);
            return Math.max(leftMax, rightMax);
        }
    }

    private int n;
    private int[] values;
    private List<Integer>[] adj;
    private int[] parent, depth, subtreeSize, heavyChild, chainId, posInChain, chainHead;
    private List<List<Integer>> chainNodes;
    private List<SegmentTree> segTrees;

    /**
     * Initialize HLD.
     *
     * @param n Number of vertices (0 to n-1)
     * @param edges List of edges (u, v)
     * @param values Vertex values
     */
    @SuppressWarnings("unchecked")
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public HeavyLightDecomposition(int n, List<int[]> edges, int[] values) {
        this.n = n;
        this.values = values.clone();
        this.adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            this.adj[i] = new ArrayList<>();
        }

        for (int[] edge : edges) {
            this.adj[edge[0]].add(edge[1]);
            this.adj[edge[1]].add(edge[0]);
        }

        // Initialize HLD arrays
        this.parent = new int[n];
        this.depth = new int[n];
        this.subtreeSize = new int[n];
        this.heavyChild = new int[n];
        this.chainId = new int[n];
        this.posInChain = new int[n];
        this.chainHead = new int[n];
        this.chainNodes = new ArrayList<>();
        this.segTrees = new ArrayList<>();

        Arrays.fill(heavyChild, -1);
        Arrays.fill(chainId, -1);
        Arrays.fill(posInChain, -1);
        Arrays.fill(chainHead, -1);

        if (n > 0) {
            dfs1(0, -1, 0);
            dfs2(0, 0);
        }
    }

    private void dfs1(int u, int p, int d) {
        parent[u] = p;
        depth[u] = d;
        subtreeSize[u] = 1;

        for (int v : adj[u]) {
            if (v != p) {
                dfs1(v, u, d + 1);
                subtreeSize[u] += subtreeSize[v];

                if (heavyChild[u] == -1 || subtreeSize[v] > subtreeSize[heavyChild[u]]) {
                    heavyChild[u] = v;
                }
            }
        }
    }

    private void dfs2(int u, int cid) {
        chainId[u] = cid;

        while (cid >= chainNodes.size()) {
            chainNodes.add(new ArrayList<>());
        }

        if (chainNodes.get(cid).isEmpty()) {
            chainHead[cid] = u;
        }

        posInChain[u] = chainNodes.get(cid).size();
        chainNodes.get(cid).add(u);

        if (heavyChild[u] != -1) {
            dfs2(heavyChild[u], cid);
        }

        for (int v : adj[u]) {
            if (v != parent[u] && v != heavyChild[u]) {
                dfs2(v, chainNodes.size());
            }
        }
    }

    private void buildSegmentTrees() {
        segTrees.clear();
        for (List<Integer> chain : chainNodes) {
            int[] chainValues = new int[chain.size()];
            for (int i = 0; i < chain.size(); i++) {
                chainValues[i] = values[chain.get(i)];
            }
            segTrees.add(new SegmentTree(chainValues));
        }
    }

    /**
     * Update value at vertex u.
     *
     * @param u Vertex index
     * @param val New value
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public void update(int u, int val) {
        values[u] = val;
        buildSegmentTrees();
    }

    /**
     * Query max value on path from u to v.
     *
     * @param u Start vertex
     * @param v End vertex
     * @return Maximum value on path u-v
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public int queryPath(int u, int v) {
        if (segTrees.isEmpty()) {
            buildSegmentTrees();
        }

        int result = 0;

        // Bring u and v to same level
        while (depth[u] > depth[v]) {
            result = Math.max(result, queryUp(u));
            u = parent[u];
        }
        while (depth[v] > depth[u]) {
            result = Math.max(result, queryUp(v));
            v = parent[v];
        }

        // Move both up simultaneously
        while (u != v) {
            result = Math.max(result, queryUp(u));
            result = Math.max(result, queryUp(v));
            u = parent[u];
            v = parent[v];
        }

        result = Math.max(result, values[u]);
        return result;
    }

    private int queryUp(int u) {
        int cid = chainId[u];
        int start = posInChain[chainHead[cid]];
        int end = posInChain[u];
        return segTrees.get(cid).query(start, end);
    }

    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        // Example: Tree with path queries
        //     0
        //    / \
        //   1   2
        //  / \
        // 3   4

        int n = 5;
        List<int[]> edges = Arrays.asList(
                new int[]{0, 1}, new int[]{0, 2},
                new int[]{1, 3}, new int[]{1, 4}
        );
        int[] values = {10, 20, 30, 40, 50};

        HeavyLightDecomposition hld = new HeavyLightDecomposition(n, edges, values);

        System.out.println("Path 3->4 max value: " + hld.queryPath(3, 4));  // 40, 20
        System.out.println("Path 4->5 max value: " + hld.queryPath(4, 4));  // 50
        System.out.println("Path 2->4 max value: " + hld.queryPath(2, 4));  // 30, 10, 20, 50

        hld.update(1, 100);
        System.out.println("After update(1, 100):");
        System.out.println("Path 3->4 max value: " + hld.queryPath(3, 4));  // 40, 100
    }
}
